pkg/services/hcloud/server/server.go

```diff
@@ -92,42 +105,97 @@ func (s *Service) Reconcile(ctx context.Context) (res reconcile.Result, err erro
 
 	conditions.MarkTrue(s.scope.HCloudMachine, infrav1.BootstrapReadyCondition)
 
-	// try to find an existing server
-	server, err := s.findServer(ctx)
-	if err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed to get server: %w", err)
+	var server *hcloud.Server
+	if s.scope.HCloudMachine.Spec.ProviderID != nil {
+		server, err = s.findServer(ctx)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("findServer: %w", err)
+		}
+
+		if server == nil {
+			// The server did disappear in HCloud? Maybe it was delete via web-UI.
+			// We set MachineError. CAPI will delete machine.
+			msg := fmt.Sprintf("hcloud server (%q) no longer available. Setting MachineError.",
+				*s.scope.HCloudMachine.Spec.ProviderID)
+
+			ctrl.LoggerFrom(ctx).Error(errors.New(msg), msg,
+				"ProviderID", *s.scope.HCloudMachine.Spec.ProviderID,
+				"BootState", s.scope.HCloudMachine.Status.BootState,
+				"BootStateSince", s.scope.HCloudMachine.Status.BootStateSince,
+			)
+
+			s.scope.SetError(msg, capierrors.CreateMachineError)
+			s.scope.HCloudMachine.SetBootState(infrav1.HCloudBootStateUnset)
+			record.Warn(s.scope.HCloudMachine, "NoHCloudServerFound", msg)
+
+			// no need to requeue.
+			return reconcile.Result{}, nil
+		}
+		updateHCloudMachineStatusFromServer(s.scope.HCloudMachine, server)
 	}
 
-	// if no server is found we have to create one
-	if server == nil {
-		server, err = s.createServer(ctx)
-		if err != nil {
-			if errors.Is(err, errServerCreateNotPossible) {
-				return reconcile.Result{RequeueAfter: 5 * time.Minute}, nil
-			}
-			return reconcile.Result{}, fmt.Errorf("failed to create server: %w", err)
+	switch s.scope.HCloudMachine.Status.BootState {
+	case infrav1.HCloudBootStateUnset:
+		return s.handleBootStateUnset(ctx)
+	case infrav1.HCloudBootStateBootToRealOS:
+		return s.handleBootToRealOS(ctx, server)
+	case infrav1.HCloudBootStateOperatingSystemRunning:
+		return s.handleOperatingSystemRunning(ctx, server)
+	default:
+		return reconcile.Result{}, fmt.Errorf("unknown BootState: %s", s.scope.HCloudMachine.Status.BootState)
+	}
+}
+
+func (s *Service) handleBootStateUnset(ctx context.Context) (reconcile.Result, error) {
+	m := s.scope.HCloudMachine
+
+	if m.Spec.ProviderID != nil && *m.Spec.ProviderID != "" {
+		// This machine seems to be an existing machine which was created before introducing
+		// Status.BootState.
+
+		if !m.Status.Ready {
+			m.SetBootState(infrav1.HCloudBootStateBootToRealOS)
+			ctrl.LoggerFrom(ctx).Info("Updating old resource (pre BootState)",
+				"to", m.Status.BootState)
+			return reconcile.Result{RequeueAfter: requeueImmediately}, nil
 		}
+
+		m.SetBootState(infrav1.HCloudBootStateOperatingSystemRunning)
+		ctrl.LoggerFrom(ctx).Info("Updating old resource (pre BootState)",
+			"to", m.Status.BootState)
+
+		// Requeue once the new way. But in most cases nothing should have changed.
```

Comment:
what do you mean by this? And you just set it on "running", what else is supposed to happen?